lib/internal/modules/esm/default_resolve.js

```diff
@@ -110,10 +110,14 @@ function resolve(specifier, parentURL) {
   if (ext === '.js' || (!format && isMain))
     format = getPackageType(url.href) === TYPE_MODULE ? 'module' : 'commonjs';
   if (!format) {
-    if (esModuleSpecifierResolution === 'node')
+    if (experimentalSpeciferResolution === 'node') {
```

Comment:
should we emit the warning on the use of the flag or only when it is active?

Reply:
I tend to prefer the pattern in general of having the warning on first use of the feature. That's what we're looking at doing for the other experimental modules features.

Reply:
The only problem is this might not catch all usage though - because we have branches in the C++ resolver which check this which may not still catch this path (this path is specifically for non JS file extensions).

So we should either bring the warning to the C++ code, or we should make it a general warning on startup.

Reply:
Also, @pd4d10 has been working on C++ warnings for these flags in https://github.com/nodejs/node/pull/30617. The approach there could apply to this PR, depending on which merges first.

Reply:
Is getting the warning into C++ layer blocking this from landing or should we do it in a follow up?